papis/commands/update.py

```diff
@@ -39,25 +107,237 @@
+        )
+    except papis.format.FormatFailedError:
+        logger.debug("Value '%s' of key '%s' isn't a papis format string.", value, key)
+    return value
+
+
+def try_parsing_str(key: str, value: str) -> str:
+    """
+    Tries to parse the input string as a python expression.
+
+    :returns: The parsed input string if string is a python expression,
+        otherwise an unchanged string.
+    """
+    try:
+        # TODO: elsewhere in the codeabase `eval()` seems to be used
+        # but this seems saver? Something for another PR?
+        value = ast.literal_eval(value)
+    except (SyntaxError, ValueError):
+        logger.debug("Value '%s' of key '%s' is not a python expression.", value, key)
+    return value
+
+
+def is_value_admissible(key: str, value: Any) -> bool:
+    """
+    Checks if the type of the input `value` is what is expected for values of `key`.
+
+    :returns: True if the value is admissible, False if it isn't.
+    """
+    success = True
+    if key in KEY_TYPES and KEY_TYPES[key] != type(value):
+        logger.error(
+            "We got the value '%s' which is of type '%s' but keys '%s' require \
+            type '%s'. No changes have been applied. Aborting...",
+            value,
+            type(value).__name__,
+            key,
+            KEY_TYPES[key].__name__,
+        )
+        success = False
+    return success
+
+
+def run_set(
+    document: papis.document.DocumentLike, to_set: List[Tuple[str, str]]
+) -> Tuple[dict[str, Any], bool]:
+    """
+    Processes a list of `to_set` tuples and applies the resulting changes to the
+    input document.
+
+    :returns: A tuple of the updated document and a boolean indicating whether
+        the update was successful.
+    """
+    processed_tuples: Dict[str, Any] = {}
+    success = True
+    for key, value in to_set:
+        processed_value: Any = try_parsing_str(key, value)
+        print(value)
+        # TODO: I'm aborting if an error is found rather than processing whatever is
+        #       correct. I noticed this is different from some other places in papis
+        #       (e.g. `doctor`). But I think it's better to as obvious as possible when
+        #       stuff is wrong. Let me know if I should change it (same applies to other
+        #       functions).
+        if not is_value_admissible(key, processed_value):
+            processed_tuples = {}
+            success = False
+            break
+        if isinstance(processed_value, str):
+            processed_value = try_formatting_str(key, value, document)
+        if key == "notes":
+            processed_value = papis.utils.clean_document_name(value)
+        if key == "files":
+            files = value
+            processed_value = list()
+            for file in files:
+                processed_value.append(papis.utils.clean_document_name(file))
+
+        processed_tuples[key] = processed_value
+
+    document.update(processed_tuples)
+    return document, success
```

Comment:
`Document` (and `dict`) are mutable collections, so you don't actually need to return it here because the input `document` will get updated.

Actually for all of these functions, maybe it would be better to not do the `document.update` part and just return a `Dict[str, Any]` with the parsed keys?

Reply:
The only problem here is with deleting keys, which can't be nicely represented in a `Dict[str, Any]`. Although I guess we could set them to empty values or None and then remove the keys with such values?

Reply:
Yeah, you're right, that's too much trouble!